python/kastore/store.py

```diff
@@ -290,13 +283,20 @@ def _read_file(self):
             raise exceptions.VersionTooNewError()
         num_items, file_size = struct.unpack("<IQ", header[12:24])
         logger.debug("Loading {} items from {} bytes".format(num_items, file_size))
-        if self._file_size != file_size:
+        if file_size < HEADER_SIZE:
             raise exceptions.FileFormatError("Bad file size in header")
+        if num_items == 0:
+            raise exceptions.FileFormatError("Empty kastore. How boring")
+        return num_items, file_size
+
+    def _read_descriptors(self, num_items, file_size):
         descriptor_block_size = num_items * ItemDescriptor.size
-        descriptor_block = self._buffer[HEADER_SIZE: HEADER_SIZE + descriptor_block_size]
+        if (HEADER_SIZE + descriptor_block_size > file_size):
+            raise exceptions.FileFormatError("Bad file size in header")
 
-        if len(descriptor_block) != descriptor_block_size:
-            raise exceptions.FileFormatError("Truncated file")
+        descriptor_block = self._file.read(descriptor_block_size)
+        if len(descriptor_block) < descriptor_block_size:
+            raise EOFError("Unexpected end of file")
```

Comment:
My understanding is that when you have a pipe (or socket, or whatever), then you can get EOF if the writer end `close()`s (the nominal case here), or if the writer process is killed, or otherwise disconnected. I'll do some reading to double check my assumptions this - I'm a bit rusty here.

Reply:
I'm thinking about this from a consumer perspective - you'll want to know the difference between the nominal case you've-read-n-stores-and-exited-gracefully vs you've-tried-to-read-a-store-and-got-truncated-input. Otherwise, you won't be able to tell if there was an error reading the n+1'th store, and you'll mistakenly think that you've correctly read all the input. So, if there *is* a network error or whatever, the last partially read store will be silently dropped and you've no way of telling this happened. Whereas if we return a TruncatedInputError instead of EOF, you will.

See what I'm getting at?

Reply:
But what if the connection gets dropped between when the writer writes the n'th and n+1'th store? It will return EOF then, but really the input is truncated. I'm not sure that this is identifiable in general.

I expect that the semantics of this stuff will be clearer once we look at some working code which uses the `openf()` case. How about I tackle that next, and we revisit this. If that doesn't clear anything up, I'm happy to go along with however you want it.

Reply:
OK, let's do that.

Reply:
I've opened  #96 to track.